### DISCUSSION TITLE
Comparing Build Ids or Tags is not working

### DISCUSSION THREAD

**Discussion by Quallos** [rfennell/AzurePipelines#1543] [Q&A]
Hello!
I wanted to have the set the Ids of specific Builds (which are not on a release pipeline right now) and compare them against each other and then create release notes out of that. This is for testing purposes. 

What I am missing is:
- Only commits are show from the latest build
- Workitems are only shown, when they are new on the current build 

What I understood from the setup is that I have different options to compare specific builds.

Please correct me if I am wrong!

(Version 1) However I am not sure if I am using the correct Ids or setup. Maybe you can help me with that.

          - task: XplatGenerateReleaseNotes@4
            displayName: 'Generate Release Notes'
            inputs:
              outputfile: '$(Build.BuildNumber).md'
              templateLocation: 'File'
              templatefile: 'xxxx/build-handlebars-template.md'
              dumpPayloadToConsole: false
              dumpPayloadToFile: false
              replaceFile: True
              appendToFile: True
              getParentsAndChildren: False
              getAllParents: False
              getIndirectPullRequests: False
              stopOnError: False
              considerPartiallySuccessfulReleases: False
              checkForManuallyLinkedWI: False
              wiqlFromTarget: WorkItems
              checkstages: true
              overrideStageName: the number of a previous " $(Build.BuildNumber)" 

(Version 2) I also tried to use diffrent Tags, this is how my setup looked like

          - task: XplatGenerateReleaseNotes@4
            displayName: 'Generate Release Notes'
            inputs:
              outputfile: '$(Build.BuildNumber).md'
              templateLocation: 'File'
              templatefile: 'xxxx/build-handlebars-template.md'
              dumpPayloadToConsole: false
              dumpPayloadToFile: false
              replaceFile: True
              appendToFile: True
              getParentsAndChildren: False
              getAllParents: False
              getIndirectPullRequests: False
              stopOnError: False
              considerPartiallySuccessfulReleases: False
              checkForManuallyLinkedWI: False
              wiqlFromTarget: WorkItems
              checkstages: true
              tags: '1.3.0'

Log for Version 1 (with BuildIds)

Get any manual test run details associated with the build
##[debug]Get batch of manual test runs [0] - [1000]
##[debug]Found 0 manual test runs associated with the build
##[debug]No manual test plans associated with build
Removing duplicate Commits from master list
Expanding the truncated commit messages...
Expanded truncated commit messages 0
Leaving CS in default order as returned by API
Find any WorkItems linked from GitHub using the AB#123 format
Adding 0 found using AB#123 links in GitHub comments
Removing duplicate WorkItems from master list
Getting tests linked to WorkItems
Checking for user defined WIQL query
No WIQL query parameters defined
Leaving WI in default order as returned by API
##[debug]Getting the details of build 7338841 from default project
The default artifact for the release was not linked to an Azure DevOps Git Repo Pull Request
Getting all completed Azure DevOps Git Repo PRs in the Team Project IngOne
Getting associated PRs
##[debug]Get batch of PRs [0] - [1000]
##[debug]Adding batch of 1 PRs
Found 1 Azure DevOps PRs in the repo
Checking for PRs associated with the commit ba33e4145870142e241c65ce6ea4fd7825dd1b84
Enriching known Pull Requests
Total Builds: [0]
Total Commits: [1]
Total Workitems: [0]
Total Related Workitems (Parent/Children): [0]
Total Release Tests: [0]
Total Tests: [0]
Total Manual Test Runs: [0]

Log for Version 2 (with Tags)

Removing duplicate Commits from master list
Expanding the truncated commit messages...
Expanded truncated commit messages 0
Leaving CS in default order as returned by API
Find any WorkItems linked from GitHub using the AB#123 format
Adding 0 found using AB#123 links in GitHub comments
Removing duplicate WorkItems from master list
Getting tests linked to WorkItems
Checking for user defined WIQL query
No WIQL query parameters defined
Leaving WI in default order as returned by API
##[debug]Getting the details of build 7338806 from default project
The default artifact for the release was not linked to an Azure DevOps Git Repo Pull Request
Getting all completed Azure DevOps Git Repo PRs in the Team Project IngOne
Getting associated PRs
##[debug]Get batch of PRs [0] - [1000]
##[debug]Adding batch of 1 PRs
Found 1 Azure DevOps PRs in the repo
Checking for PRs associated with the commit 6caee70d359e6a687c49036e063e9302e5d777a0
Checking for PRs associated with the commit 172bb39e255854ed2bd13ac535ad2cc6085041dc
Checking for PRs associated with the commit 760e1409e4a44e5e7650fda49560ce5be762f190
Enriching known Pull Requests
Total Builds: [0]
Total Commits: [3]
Total Workitems: [0]
Total Related Workitems (Parent/Children): [0]
Total Release Tests: [0]
Total Tests: [0]
Total Manual Test Runs: [0]
Total Manual Test Configurations: [0]
Total Pull Requests: [0]
Total Indirect Pull Requests: [0]
Total Associated Test WI: [0]
Total Published Artifacts: [0]
Total Consumed Artifacts: [0]
Total WIQL Workitems: [0]
Generating the release notes, the are 1 template(s) to process
##[debug]Using template mode File

**Comment by rfennell** [comment]
The task finds CS/WI associated with all runs since the last successful build of a given build/pipeline. So if each of your runs don't error the task only finds the changes of the current run. 

But it can do more than that with Multistage YAML or Classic Releases, as detailed in the [WIKI](https://github.com/rfennell/AzurePipelines/wiki/GenerateReleaseNotes---Node-based-Cross-Platform-Task


## Use Cases
> The a video on usage of this task is discussed in this [DDD Community Conference Session](https://www.youtube.com/watch?v=xaV3dFoQdV8&t=331s)

There are various ways that the task can be used. A common pattern is to use the task multiple times in a CI/CD pipeline.

1. Run once for every CI build, so you get the build notes of what is new in that build
1. Run as part of a CD release stage, checking against the last successful deployment to that stage, to generate the release notes specific to that release to a given environment.

Possible sets of parameters depending on your usage are summarized below

| Option | Multi Stage YAML | Classic Build/Release |
|-|-|-|
| Generate notes for just the current build | Requires `checkstages=false` parameter| Run inside the build |
| Generate notes since the last successful release.  <br>Option 1. Place the task in a stage that is only run when you wish to generate release notes. Usually this will be guarded by branch based filters or manual approvals.  |  Requires `checkstages=true` parameter | Run inside the release. Supported and you can override the stage name used for comparison using the `overrideStageName` parameter
| Generate notes since the last successful release.  <br>Option 2. Set the task to look back for the last successful build that has a given tag |  Requires `checkstages=true` and the `tags` parameters| Not supported
| Generate notes since the last successful release.  <br>Option 3. Override the build that the task uses for comparison with a fixed value |  Requires `checkstages=true` and the `overrideBuildReleaseId` parameters | Run inside the release. Requires the `overrideBuildReleaseId` parameter

## Where do WI and Changesets come from?
At it's core this task uses the Azure DevOps REST API to get the list of work items and commits/changesets either directly associated with a build, or between two builds/releases (see above). 

However, there are also other means to supplement these lists by setting the following parameters

- To get parent and child work items of those directly associated with the build/release set `getAllParents` and `getParentsAndChildren ` parameters
- To get manually linked work items set the `checkForManuallyLinkedWI` parameter
- To match commits to Azure DevOps PR cross project within the organisation set the `searchCrossProjectForPRs` parameter
- To use a WIQL query to get additional work items set the `wiqlWhereClause` and `wiqlFromTarget` parameters or the `wiqlSharedQueryName` parameter
